src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -0,0 +1,244 @@
+        }
+
+        private void SetSolutionProperties(string solutionFilePath)
+        {
+            const string SolutionDirProperty = "SolutionDir";
+
+            // When MSBuild is building an individual project, it doesn't define $(SolutionDir).
+            // However when building an .sln file, or when working inside Visual Studio,
+            // $(SolutionDir) is defined to be the directory where the .sln file is located.
+            // Some projects out there rely on $(SolutionDir) being set (although the best practice is to
+            // use MSBuildProjectDirectory which is always defined).
+            if (!string.IsNullOrEmpty(solutionFilePath))
+            {
+                var solutionDirectory = Path.GetDirectoryName(solutionFilePath);
+                if (!solutionDirectory.EndsWith(@"\", StringComparison.Ordinal))
+                {
+                    solutionDirectory += @"\";
+                }
+
+                if (Directory.Exists(solutionDirectory))
+                {
+                    _properties = _properties.SetItem(SolutionDirProperty, solutionDirectory);
+                }
+            }
+        }
+
+        private DiagnosticReportingMode GetReportingModeForUnrecognizedProjects()
+            => this.SkipUnrecognizedProjects
+                ? DiagnosticReportingMode.Log
+                : DiagnosticReportingMode.Throw;
+
+        /// <summary>
+        /// Loads the <see cref="SolutionInfo"/> for the specified solution file, including all projects referenced by the solution file and 
+        /// all the projects referenced by the project files.
+        /// </summary>
+        public async Task<SolutionInfo> LoadSolutionInfoAsync(
+            string solutionFilePath,
+            IProgress<ProjectLoadProgress> progress = null,
+            CancellationToken cancellationToken = default)
+        {
+            if (solutionFilePath == null)
+            {
+                throw new ArgumentNullException(nameof(solutionFilePath));
+            }
+
+            if (!_pathResolver.TryGetAbsoluteSolutionPath(solutionFilePath, baseDirectory: Directory.GetCurrentDirectory(), DiagnosticReportingMode.Throw, out var absoluteSolutionPath))
+            {
+                // TryGetAbsoluteSolutionPath should throw before we get here.
+                return null;
+            }
+
+            using (_dataGuard.DisposableWait(cancellationToken))
+            {
+                this.SetSolutionProperties(absoluteSolutionPath);
+            }
+
+            var solutionFile = MSB.Construction.SolutionFile.Parse(absoluteSolutionPath);
+
+            var reportingMode = GetReportingModeForUnrecognizedProjects();
+
+            var requestedProjectOptions = new DiagnosticReportingOptions(
+                onPathFailure: reportingMode,
+                onLoaderFailure: reportingMode);
+
+            var discoveredProjectOptions = new DiagnosticReportingOptions(
```

Comment:
>discoveredProjectOptions [](start = 16, length = 24)

Want to check that you deliberately created these two locals with exact same options.

Reply:
yes, that's right. For the solution case, they have the same options. I'll add a comment.

Reply:
Instead of a comment, I made it clearer that this is intentional by creating one set of options and then passing it to both parameters while naming the arguments.